Index: linux-2.6.36/drivers/net/ppp_generic.c
===================================================================
--- linux-2.6.36.orig/drivers/net/ppp_generic.c	2017-12-05 20:23:24.094020054 +0800
+++ linux-2.6.36/drivers/net/ppp_generic.c	2017-12-05 20:42:43.562131400 +0800
@@ -2644,6 +2644,11 @@
       num[2] = u<<(32-(FIELD2))>>(32-FIELD2);
       if (num[1] == 0)
       {
+      #if defined(CONFIG_ZYXEL_PPTP_SUPPORT)/* __ZyXEL__, MelissaChen, 20171122, Support PPTP VPN.  */
+         if (num[2] == 11 ){
+            sprintf(dev->name, "ppp%d", num[2]);
+         }else
+      #endif
          sprintf(dev->name, "ppp%d", num[0]);
       }
       else if (num[1] == 1) /* vlan mux */
Index: linux-2.6.36/net/netfilter/Kconfig
===================================================================
--- linux-2.6.36.orig/net/netfilter/Kconfig	2017-12-05 20:23:24.394020053 +0800
+++ linux-2.6.36/net/netfilter/Kconfig	2017-12-05 20:35:13.787131400 +0800
@@ -1089,6 +1089,9 @@
 config ZYXEL_NF_SESSION_CTL
         bool "ZyXEL feature - netfilter session control"
 
+config ZYXEL_PPTP_SUPPORT
+        bool "Support PPTP VPN"
+
 endmenu
 
 source "net/netfilter/ipvs/Kconfig"
