Index: qtn_image/source/buildroot/package/zhostapd/hostapd-2.1/src/ap/wps_hostapd.c
===================================================================
--- qtn_image.orig/source/buildroot/package/zhostapd/hostapd-2.1/src/ap/wps_hostapd.c	2017-01-13 16:14:19.000000000 +0800
+++ qtn_image/source/buildroot/package/zhostapd/hostapd-2.1/src/ap/wps_hostapd.c	2017-01-24 10:27:25.761616609 +0800
@@ -729,7 +729,9 @@
 			       ifname);
 
 	wpa_printf(MSG_DEBUG, "WPS: AP configuration updated");
-
+#if 1 //ZYXEL,Mark	
+	hostapd_send_wlan_msg(hapd, "WPS AP configuration updated");
+#endif
 	return 0;
 }
 
